File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.py
def find_row_and_column(quotient, mod, n):
    if mod == 0:
        x = quotient-1
        if quotient % 2 == 0:
            y = 0
        else:
            y = n - 1
    else:
        x = quotient
        if quotient % 2 == 0:
            y = mod - 1
        else:
            y = n - mod
    return (x, y)

def solution(n, w, num):
    answer = 0
    
    quotient = num // w # num의 몫
    mod = num % w # num의 나머지
    x, y = find_row_and_column(quotient, mod, w) # num이 위치하는 행, 열
    
    last_x, last_y = find_row_and_column(n//w, n%w, w) # 마지막 상자의 행, 열
    if last_x % 2 == 0:
        if last_y < y:
            answer = last_x - x
        else:
            answer = last_x - x + 1
    else:
        if last_y <= y:
            answer = last_x - x + 1
        else:
            answer = last_x - x
    
    return answer

File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.java
class Solution {
    public int solution(int n, int w, int num) {
        int answer = 1;
        
        int quotient = num / w;
        int mod = num % w;
        int[] xy = findRowAndColumn(quotient, mod, w);
        int x = xy[0], y = xy[1]; // 꺼내려는 상자가 위치한 행과 열
        
        int[] lastXY = findRowAndColumn(n / w, n % w, w);
        int lastX = lastXY[0], lastY = lastXY[1]; // 마지막 상자가 위치한 행과 열
        
        if (lastX % 2 == 0) {
            if (lastY < y) {
                answer += lastX - x - 1;
            } else {
                answer += lastX - x;
            }
        } else {
            if (lastY <= y) {
                answer += lastX - x;
            } else {
                answer += lastX - x - 1;
            }
        }
        
        return answer;
    }
    
    // 특정 상자가 위치한 행, 열 구하는 메서드
    static int[] findRowAndColumn(int quotient, int mod, int w) {
        int[] xy = new int[2];
        
        if (mod == 0) {
            xy[0] = quotient - 1;
            if (quotient % 2 != 0) xy[1] = w - 1;
        } else {
            xy[0] = quotient;
            if (quotient % 2 == 0) xy[1] = mod - 1;
            else xy[1] = w - mod;
        }
        
        return xy;
    }
}